text/add-bf16-f64f64-and-f80-type.md

```diff
@@ -0,0 +1,121 @@
+- Feature Name: `add-bf16-f64f64-and-f80-type`
+- Start Date: 2023-7-10
+- RFC PR: [rust-lang/rfcs#3456](https://github.com/rust-lang/rfcs/pull/3456)
+- Rust Issue: [rust-lang/rust#2629](https://github.com/rust-lang/rfcs/issues/2629)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes new floating point types to enhance FFI with specific targets:
+
+- `bf16` as builtin type for the 'brain floating point' format, widely used in machine learning, different from the IEEE 754 standard `binary16` representation
+- `f64f64` in `core::arch` for the legacy extended float format used in the PowerPC architecture
+- `f80` in `core::arch` for the extended float format used in the x86 and x86_64 architectures
+
+Also, this proposal introduces `c_longdouble` in `core::ffi` to represent the correct format for 'long double' in C.
+
+# Motivation
+[motivation]: #motivation
+
+The types listed above may be widely used in existing native code, but are not available on all targets. Their underlying representations are quite different from 16-bit and 128-bit binary floating format defined in IEEE 754.
+
+In respective targets (namely PowerPC and x86), the target-specific extended types are referenced by `long double`, which makes `long double` ambiguous in the context of FFI. Thus defining `c_longdouble` should help interoperating with C code using the `long double` type.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`bf16` is available on all targets. The operators and constants defined for `f32` are also available for `bf16`.
+
+For `f64f64` and `f80`, their availability is limited to the following targets, but this may change over time:
+
+- `f64f64` is supported on `powerpc-*` and `powerpc64(le)-*`, available in `core::arch::{powerpc, powerpc64}`
+- `f80` is supported on `i[356]86-*` and `x86_64-*`, available in `core::arch::{x86, x86_64}`
+
+The operators and constants defined for `f32` and `f64` are available for `f64f64` and `f80` in their respective arch-specific modules.
+
+All proposed types do not have literal representation. Instead, they can be converted to or from IEEE 754 compliant types.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `bf16` type
+
+`bf16` consists of 1 sign bit, 8 bits of exponent, 7 bits of mantissa. Some ARM, AArch64, x86 and x86_64 targets support `bf16` operations natively. For other targets, they will be promoted into `f32` before computation and truncated back into `bf16`.
+
+`bf16` will generate the `bfloat` type in LLVM IR.
+
+## `f64f64` type
+
+`f64f64` is the legacy extended floating point format used on PowerPC targets. It consists of two `f64`s, with the former acting as a normal `f64` and the latter for an extended mantissa.
+
+The following `From` traits are implemented in `core::arch::{powerpc, powerpc64}` for conversion between `f64f64` and other floating point types:
```

Comment:
It sounds like this type does not have semantics that are equivalent to any IEEE float type. But we need some document to explain exactly what their semantics are, i.e. the exact bits you get out when doing arithmetic on values of this type. Does such a document exist?

Reply:
That's a good question. Basically, it would act like an f64x2 vector, but when exactly bits will move between the two is a question that would need to be answered.

Reply:
> For the most part, it would act like an f64x2 vector

wait, that's not at all how `f64f64` arithmetic works, it instead works more like a big-integer. e.g. here's multiplying two double-double values in the `twofloat` crate: https://docs.rs/twofloat/0.7.0/src/twofloat/arithmetic.rs.html#145

Reply:
Yeah that's more like what I expected. Is there an explanation somewhere of what the "meaning" of such an `(a, b)` pair is, i.e. what is its mathematical-valued semantics? Is it `a + b` (where this is mathematical inf-precision `+` on rational numbers)?

Is the behavior of all basic operations on that kind of representation exactly guaranteed, the same way IEEE exactly guarantees behavior for our regular floats?

Reply:
Not knowing the details of the `fast_two_sum` function, that looks to be a binomial product, which is what I was referring to with "multiplication/division, etc., would cross" though I guess I wasn't quite clear on that.

Reply:
> Given the unusual semantics and the somewhat legacy nature of `f64f64`, would it be better to just provide a type with no methods/trait implementations

sounds good to me! though I'd at least have `Copy`, `Clone`, `Default`, and `Debug`, where `Debug` could just be as if it was: `struct f64f64 { high: f64, low: f64 }`

Reply:
> Not knowing the details of the `fast_two_sum` function, that looks to be a binomial product, which is what I was referring to with "multiplication/division, etc., would cross" though I guess I wasn't quite clear on that.

ok, yeah. addition and subtraction also don't act like a `f64x2`, about the only ops that act like `f64x2` are neg, abs, and copy.

Reply:
> Not knowing the details of the `fast_two_sum` function, that looks to be a binomial product, which is what I was referring to with "multiplication/division, etc., would cross" though I guess I wasn't quite clear on that.

So -- what is the mathematical value of a pair `(a, b)` then, the rational number this represents?

Reply:
> Yeah that's more like what I expected. Is there an explanation somewhere of what the "meaning" of such an `(a, b)` pair is, i.e. what is its mathematical-valued semantics? Is it `a + b` (where this is mathematical inf-precision `+` on rational numbers)?

yee, it is `high + low` where the number is the exact mathematical sum of two `f64`s

> Is the behavior of all basic operations on that kind of representation exactly guaranteed, the same way IEEE exactly guarantees behavior for our regular floats?

I've heard that many special functions (like sin, cos, etc.) don't even always return canonical values (as in the result is represented differently than the exact same number would be by arithmetic ops), idk which ones.